File: dropwizard-kafka/src/main/scala/com/datasift/dropwizard/kafka/config/KafkaClientConfiguration.scala
package com.datasift.dropwizard.kafka.config

import com.yammer.dropwizard.config.Configuration
import reflect.BeanProperty
import javax.validation.Valid
import javax.validation.constraints.NotNull
import com.datasift.dropwizard.zookeeper.config.ZooKeeperConfiguration
import com.yammer.dropwizard.util.Duration

/** [[com.yammer.dropwizard.config.Configuration]] for connecting to a Kafka cluster */
class KafkaClientConfiguration extends Configuration {

  /** configuration for the cluster's ZooKeeper quroum */
  @BeanProperty
  @NotNull
  @Valid
  val zookeeper = new ZooKeeperConfiguration

  /** client socket timeout, in milliseconds */
  @BeanProperty
  @NotNull
  val socketTimeout = Duration.seconds(30)

  /** producer-specific configuration */
  @BeanProperty
  @Valid
  @NotNull
  val producer = new KafkaProducerConfiguration

  /** consumer-specific configuration */
  @BeanProperty
  @Valid
  @NotNull
  val consumer = new KafkaConsumerConfiguration
}

File: dropwizard-extra-kafka/src/main/java/com/datasift/dropwizard/kafka/config/KafkaClientConfiguration.java
package com.datasift.dropwizard.kafka.config;

import com.datasift.dropwizard.zookeeper.config.ZooKeeperConfiguration;
import com.codahale.dropwizard.util.Duration;
import com.fasterxml.jackson.annotation.JsonProperty;

import javax.validation.Valid;
import javax.validation.constraints.NotNull;

/**
 * Base configuration for Kafka clients.
 *
 * @see KafkaConsumerConfiguration
 * @see KafkaProducerConfiguration
 */
abstract public class KafkaClientConfiguration {

    /**
     * The {@link ZooKeeperConfiguration} of the ZooKeeper quorum to use.
     */
    @JsonProperty
    @Valid
    @NotNull
    protected ZooKeeperConfiguration zookeeper = new ZooKeeperConfiguration();

    /**
     * The time to wait on a network socket before timing out requests.
     */
    @JsonProperty
    @NotNull
    protected Duration socketTimeout = Duration.seconds(30);

    /**
     * @see KafkaClientConfiguration#zookeeper
     */
    public ZooKeeperConfiguration getZookeeper() {
        return zookeeper;
    }

    /**
     * @see KafkaClientConfiguration#socketTimeout
     */
    public Duration getSocketTimeout() {
        return socketTimeout;
    }
}
